crypto/rsa/rsa_lib.c

```diff
@@ -240,6 +241,56 @@ int RSA_set0_crt_params(RSA *r, BIGNUM *dmp1, BIGNUM *dmq1, BIGNUM *iqmp)
     return 1;
 }
 
+/*
+ * Is it better to export RSA_PRIME_INFO structure
+ * and related functions to let user pass a triplet?
+ */
+int RSA_set0_multi_prime_params(RSA *r, BIGNUM **primes, BIGNUM **exps,
+                                BIGNUM **coeffs, size_t pnum)
+{
+    STACK_OF(RSA_PRIME_INFO) *prime_infos;
+    RSA_PRIME_INFO *pinfo;
+    size_t i;
+
+    if (primes == NULL || exps == NULL || coeffs == NULL || !pnum)
+        return 0;
+
+    prime_infos = sk_RSA_PRIME_INFO_new_null();
+    if (prime_infos == NULL)
+        return 0;
+
+    if (r->prime_infos != NULL)
+        sk_RSA_PRIME_INFO_pop_free(r->prime_infos, rsa_multip_info_free);
+
+    for (i = 0; i < pnum; i++) {
+        pinfo = rsa_multip_info_new();
+        if (pinfo == NULL)
+            goto err;
+        if (primes[i] != NULL && exps[i] != NULL && coeffs[i] != NULL) {
+            BN_free(pinfo->r);
+            BN_free(pinfo->d);
+            BN_free(pinfo->t);
+            pinfo->r = primes[i];
+            pinfo->d = exps[i];
+            pinfo->t = coeffs[i];
+        } else {
+            rsa_multip_info_free(pinfo);
+            goto err;
+        }
+        if (!sk_RSA_PRIME_INFO_push(prime_infos, pinfo)) {
+            rsa_multip_info_free(pinfo);
+            goto err;
+        }
+    }
+
+    r->prime_infos = prime_infos;
+
+    return rsa_multip_calc_product(r);
+ err:
+    sk_RSA_PRIME_INFO_free(prime_infos);
+    return 0;
+}
```

Comment:
Any reason we can't automatically set the version number in this function?

Reply:
This is because there is no easy way to 'unset' this if the RSA structure will be reused as:
```
key = RSA_new();
/* set a bundle of primes, params for multi-prime usages */
RSA_set_version(key, 1);
/* use the multi-prime key */

/* after a while, need to use the key as a 2-prime key and reset the params and factors */
RSA_set_version(key, 0);
/* use the 2-prime key */
```
If we set the version number to 0 automatically in `RSA_set0_factors`, that would require the calling order must be: `RSA_set0_factors` then `RSA_set0_multi_prime_params`

So let the user to set version number explicitly might be a simple choice...